### PR TITLE
Update podgroups once schedulingPolicy of MPIJobs are changed

### PR DESCRIPTION
**Pull request by tenzen-y**
Previously, we implemented the logic to create and delete podGroups for the scheduler plugins in #538.

However, I forgot to implement the logic to update podGroups.

So, I implemented the logic. Also, I cleaned up the PodGroupCtrl interface so that we can control PodGroups without passing an unneeded namespace.

### CODE REVIEW

**Comment by alculquicondor** [comment]
do we have the same behavior in training-operator?

Question about coscheduling plugin:
I suppose that if the minMembers increases, but pods were already scheduled, they wouldn't be disrupted. But new pods will be blocked from scheduling until the new minMembers threshold is reached. Correct?

**Comment by tenzen-y** [comment]
> do we have the same behavior in training-operator?

Yes, the training operator updates podGroups once schedulingPolicy is updated.

https://github.com/kubeflow/common/blob/83259a047fc47895dab1012728fb829523363a6d/pkg/controller.v1/common/scheduling.go#L49-L51

> But new pods will be blocked from scheduling until the new minMembers threshold is reached. Correct?

Do new pods mean the case of increased worker replicas of MPIJob (scale out)?

File: pkg/controller/mpi_job_controller.go [hunk 1]
```diff
@@ -691,6 +692,10 @@ func (c *MPIJobController) getOrCreatePodGroups(mpiJob *kubeflow.MPIJob) (metav1
 		c.recorder.Event(mpiJob, corev1.EventTypeWarning, ErrResourceExists, msg)
 		return nil, fmt.Errorf(msg)
 	}
+
+	if !c.podGroupCtrl.pgSpecsAreEqual(podGroup, newPodGroup) {
```
Comment (alculquicondor) [inline pkg/controller/mpi_job_controller.go:696]: nit:
copy the new spec here. `podGroupCtrl` should just receive the updated object.

File: test/integration/mpi_job_controller_test.go [hunk 2]
```diff
@@ -535,6 +536,23 @@ func TestMPIJobWithSchedulerPlugins(t *testing.T) {
...
+		t.Errorf("Failed sending job to apiserver: %v", err)
+	}
+	if err = wait.Poll(waitInterval, wait.ForeverTestTimeout, func() (bool, error) {
+		pg, err := getSchedPodGroup(ctx, gangSchedulerCfg.schedClient, mpiJob)
+		if err != nil {
+			return false, err
+		}
+		return *pg.Spec.ScheduleTimeoutSeconds == updatedScheduleTimeSeconds, nil
+	}); err != nil {
+		t.Errorf("Failed updating shceduler-plugins PodGroup: %v", err)
```
Comment (alculquicondor) [inline test/integration/mpi_job_controller_test.go:554]:
```suggestion
		t.Errorf("Failed updating scheduler-plugins PodGroup: %v", err)
```

**Comment by tenzen-y** [comment]
> > Do new pods mean the case of increased worker replicas of MPIJob (scale out)?
> > yes

Correct.

File: test/integration/mpi_job_controller_test.go (hunk 2 above)
Comment (tenzen-y) [reply test/integration/mpi_job_controller_test.go:554]: Oops

**Comment by alculquicondor** [comment]
did you forget to push?

**Comment by tenzen-y** [comment]
> did you forget to push?

@alculquicondor Sorry. I forgot to push.

File: pkg/controller/mpi_job_controller.go [hunk 3]
```diff
@@ -691,6 +692,10 @@ func (c *MPIJobController) getOrCreatePodGroups(mpiJob *kubeflow.MPIJob) (metav1
 		c.recorder.Event(mpiJob, corev1.EventTypeWarning, ErrResourceExists, msg)
 		return nil, fmt.Errorf(msg)
 	}
+
+	if updatedPodGroup, equal := c.podGroupCtrl.pgSpecsAreEqual(podGroup, newPodGroup); !equal {
```
Comment (alculquicondor) [inline pkg/controller/mpi_job_controller.go:696]: oh, I see what you did and now I understand why the previous implementation was necessary: you can't update the spec unless you know the concrete type.

Sorry for the back and forth, but I think the previous implementation was better

Comment (tenzen-y) [reply pkg/controller/mpi_job_controller.go:696]: I'm ok. Let's revert this commit.

File: pkg/controller/mpi_job_controller.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/controller/mpi_job_controller.go:696]: as discussed, this is fine

**Review by alculquicondor — COMMENTED** [review]
/lgtm

**Pull request merged**